src/uct/cuda/cuda_ipc/cuda_ipc_cache.c

```diff
@@ -96,26 +166,137 @@ static void uct_cuda_ipc_cache_purge(uct_cuda_ipc_cache_t *cache)
+    if (status != UCS_OK) {
+        return status;
+    }
+
+    status =
+        UCT_CUDADRV_FUNC_LOG_ERR(cuMemAddressReserve(&dptr, key->b_len, 0, 0, 0));
+    if (status != UCS_OK) {
+        goto release_handle;
+    }
+
+    status = UCT_CUDADRV_FUNC_LOG_ERR(cuMemMap(dptr, key->b_len, 0, handle, 0));
+    if (status != UCS_OK) {
+        goto release_va_range;
     }
 
+    cuCtxGetDevice(&access_desc.location.id);
+
+    access_desc.location.type = CU_MEM_LOCATION_TYPE_DEVICE;
+    access_desc.flags         = CU_MEM_ACCESS_FLAGS_PROT_READWRITE;
+
+    status = UCT_CUDADRV_FUNC_LOG_ERR(cuMemSetAccess(dptr, key->b_len, &access_desc, 1));
+    if (status != UCS_OK) {
+        cuMemUnmap(dptr, key->b_len);
+        goto release_va_range;
+    }
+
+    *mapped_addr = dptr;
+
+    return UCT_CUDADRV_FUNC_LOG_ERR(cuMemRelease(handle));
+
+release_va_range:
+    cuMemAddressFree(dptr, key->b_len);
+release_handle:
+    cuMemRelease(handle);
+
     return status;
 }
 
+static ucs_status_t
+uct_cuda_ipc_open_memhandle_mempool(const uct_cuda_ipc_rkey_t *key,
+                                    CUdeviceptr *mapped_addr)
+{
+    CUmemoryPool *imported_mpool = NULL;
+    int key_present              = 0;
+    CUmemAccessDesc access_desc  = {};
+    CUdeviceptr dptr;
+    ucs_status_t status;
+
+    status =
+        uct_cuda_ipc_get_rem_mpool_cache((CUmemFabricHandle*)&key->ph.handle.mempool,
+                &imported_mpool, &key_present);
+    if (status != UCS_OK) {
+        return status;
+    }
+
+    ucs_assert(imported_mpool != NULL);
+
+    status = UCT_CUDADRV_FUNC_LOG_ERR(cuMemPoolImportPointer(&dptr,
+                *imported_mpool, (CUmemPoolPtrExportData*)&key->ph.ptr));
+    if (status != UCS_OK) {
+        return status;
+    }
+
+    if (!key_present) {
+        cuCtxGetDevice(&access_desc.location.id);
+        access_desc.location.type = CU_MEM_LOCATION_TYPE_DEVICE;
+        access_desc.flags         = CU_MEM_ACCESS_FLAGS_PROT_READWRITE;
+
+        status = UCT_CUDADRV_FUNC_LOG_ERR(cuMemPoolSetAccess(*imported_mpool,
+                    &access_desc, 1));
+        if (status != UCS_OK) {
+            return status;
+        }
+    }
+
+    *mapped_addr = dptr;
+
+    return UCS_OK;
+}
+#endif
+
+static ucs_status_t uct_cuda_ipc_open_memhandle(const uct_cuda_ipc_rkey_t *key,
+                                                CUdeviceptr *mapped_addr)
+{
+#if HAVE_CUDA_FABRIC
+    ucs_trace("key handle type %u", key->ph.handle_type);
+
+    if (key->ph.handle_type == UCT_CUDA_IPC_KEY_HANDLE_TYPE_LEGACY) {
+        return uct_cuda_ipc_open_memhandle_legacy(key->ph.handle.legacy, mapped_addr);
+    } else if (key->ph.handle_type == UCT_CUDA_IPC_KEY_HANDLE_TYPE_VMM) {
+        return uct_cuda_ipc_open_memhandle_vmm(key, mapped_addr);
+    } else if (key->ph.handle_type == UCT_CUDA_IPC_KEY_HANDLE_TYPE_MEMPOOL) {
+        return uct_cuda_ipc_open_memhandle_mempool(key, mapped_addr);
+    } else {
```

Comment:
can we put only the non-legacy if..else parts under the ifdef/endif scope?

Reply:
Will do